[1, 2단계 - 체스] 조조(조은별) 미션 제출합니다.

eun-byeol opened the pull request:
비밥, 안녕하세요! 조조입니다🙂

악명(?) 높은 체스 미션인 만큼 요구사항을 지키며 구현하는 것이 쉽지 않았던 것 같아요.
컨벤션 하나를 지키면 다른 것이 위배되고, 효율성과 요구사항을 지키는 것이 trade off 처럼 느껴졌습니다..!🤣

저는 24시 언제든 slack으로 응답이 가능합니다. 반대로, 비밥의 시간규칙(?)을 알려주시면 좋을 것 같아요!(ex. 몇 시 이후에는 dm 자제 등)

# 🥔 본인의 상황(feat. TMI)
> 독창적이고 신박한(?) 코드 보다는, 누구나 예측가능한 코드를 짜는 것을 목표하고 있습니다.
그러나, 스스로 판단 기준이 명확하지 않다 보니 코드를 짤 때 여러 의문점들이 생겼었습니다.
정답을 요구하는 것은 아니지만, 현직에 계신 비밥의 견해 또한 들어보고 싶습니다.

> 대학을 졸업하고 취준생의 신분에서 우테코에 들어왔습니다.
수료 후 취업을 희망하다 보니, 현직에 계신 분들의 이야기를 많이 듣고 싶습니다.
  
# 🤔 코드를 작성할 때 신경 쓴 점
## 1. 테스트 하기 쉽게 체스판 생성 로직을 만들자
테스트를 쉽게 만들기 위해 체스판 상황을 쉽게 조작할 수 있어야 했습니다.
따라서, 체스판 생성 로직을 두 개로 구분했습니다. 
1. `createInitialBoard` : 초기 체스판 생성
2. `createCustomBoard` : 커스텀 체스판 생성

다음과 같은 상황으로 체스판을 생성할 수 있습니다.
  ```java
    List<String> customBoard = List.of(
            "........",
            "...K....",
            "....Q...",
            "........",
            "........",
            "........",
            ".p......",
            ".k......"
    );

  Board.createCustomBoard(customBoard);
  ```

## 2. 상속 활용
<img width="832" alt="image" src="https://github.com/woowacourse/java-chess/assets/62530884/78ba3d2b-149d-43e3-b52b-51659dffea34">

기물들을 하나의 클래스로 관리할 수 있도록 상속을 활용했습니다.
기물이 없는 것 또한 객체로 바라보아, `None` 기물을 만들었습니다.

# 💭 질문
## 1. VO도 getter를 제한해야 하나요?
보드의 위치 좌표를 나타내는 Position VO 클래스가 있습니다.
위치 계산 로직을 위해 Position의 row와 column 값이 필요했습니다.
VO를 위한 DTO 클래스를 만들어야 하는지, VO는 getter를 허용하는지 궁금합니다.

## 2. 로직의 통일성 vs 효율성
기물들의 이동 방향 검증을 Direction enum 클래스를 활용해서 구현하고자 했었습니다.
그러나, 일부 기물(Pawn, King, Knight)들은 Direction enum을 활용하는 것보다, 내부에서 따로 계산 로직을 만드는 것이 더욱 간결했습니다.
저는 효율성을 택하여 구현하는데,
이런 상황이라면 비밥은 어떤 것을 선택하실 지 궁금합니다🤔

eun-byeol committed:
docs: 기능 목록 추가

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

masonkimseoul committed:
test: x와 y좌표에 해당하는 Position 객체를 생성한다

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

eun-byeol committed:
feat: 1 이상 8 이하의 좌표 값이 아니면 예외가 발생한다

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

eun-byeol committed:
feat: 기물 클래스 생성

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

eun-byeol committed:
refactor: Position 클래스 필드 변수명 수정

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

eun-byeol committed:
feat: 진영별 기물을 배치한다

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

eun-byeol committed:
refactor: 체스판에 초기화 로직 수정

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

masonkimseoul committed:
feat: 올바르지 않은 게임 명령어가 들어오면 예외가 발생한다

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
test: 올바른 게임 명령어가 들어오면 해당 Command 상수를 반환한다

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 게임 명령어 입력 로직 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
style: newline 추가

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 유효하지 않은 게임 명령어 재입력 로직 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

eun-byeol committed:
feat: start이면 게임을 시작하고 체스판을 출력한다

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

eun-byeol committed:
feat: end이면 게임을 종료한다

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

eun-byeol committed:
refactor: 게임 시작 메시지 출력 로직 분리

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

masonkimseoul committed:
docs: 체스 이동 기능 목록 추가

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 문자열 좌표를 받아 Position 객체를 생성한다

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 이동 시 유효하지 않은 위치면 예외가 발생한다

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 폰 1칸 전진 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 룩 상하좌우 전진 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 비숍 대각선 전진 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 퀸 상하좌우 대각선 전진 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 나이트 L자 전진 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 킹 상하좌우 대각선 1칸 전진 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 이동 시 target 위치로 이동할 수 없는 기물이면 예외가 발생한다

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 경로 상에 다른 기물 존재 확인 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 도착 좌표로 기물을 이동하고, 출발 좌표를 NONE으로 만든다

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

eun-byeol committed:
refactor: 코드 중복 제거

Co-authored-by: masonkimseoul <masonkimseoul@gmail.com>

eun-byeol committed:
refactor: LineDto 클래스명 변경

masonkimseoul committed:
feat: move이면 체스를 이동하고 결과를 출력한다

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
fix: 이동 경로 장애물 확인 로직 수정

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
refactor: Pawn canMove 로직 수정

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: Pawn 최초 이동 시 2칸 이동 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: Pawn 전방 대각선 1칸 공격 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
feat: 게임 차례 검증 기능 구현

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
style: 코드 스타일 수정

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
refactor: Direction enum 사용하여 canMove 로직 수정

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
refactor: 매직넘버 제거

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
refactor: 메서드 분리

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
refactor: 재입력 로직 수정

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

masonkimseoul committed:
style: 코드 스타일 수정

Co-authored-by: eun-byeol <joeunbyeol98@gmail.com>

pci2676 commented on src/main/java/chess/controller/ChessGame.java:
```diff
@@ -0,0 +1,71 @@
+        do {
+            board = executeGame(board);
+        } while (isRunning(board));
+    }
```
null을 다루는 코드보다 메세지를 활용하는 방식으로 개선해보시면 좋을것 같습니다.

eun-byeol replied:
controller에서 null 로직을 삭제했습니다.
- 게임 시작 전 board 초기화 선행
- 게임 상태를 객체화하여 GameStatus 클래스 생성
  - `start 입력 후 start 재입력`, `start 입력 전 move 입력`과 같은 상황에 에러를 던져 재입력 받도록 했습니다.

pci2676 commented on src/main/java/chess/model/Board.java:
```diff
@@ -0,0 +1,147 @@
+
+    public static Board createCustomBoard(List<String> customBoard) {
+        return new Board(generateBoard(customBoard), 0);
+    }
```
아직 createCustomBoard를 사용하지는 않으시는것 같습니다.

Board를 어떻게 만드는지에 대한 제어를 하고 싶으시다면 제어 하고싶은 영역의 의존성을 역전시켜보면 좋을것 같습니다.

eun-byeol replied:
결론 : `generate` 규약을 정의한 `BoardFactory` 인터페이스를 만들고, `InitialBoardFactory`, `CustomBoardFactory`에서 각각 구현하도록 만들었습니다. 따라서 이런식으로 Board를 생성할 수 있도록 수정했습니다.
- InitialBoardFactory
```java
BoardFactory boardFactory = new InitialBoardFactory();
Board board = boardFactory.generate();
```
- CustomBoardFactory
```java
List<String> snapShot = List.of(
    "........",
    "K...P...",
    "PP......",
    "........",
    "...p....",
    ".pn.....",
    ".p.q....",
    "....k..."
);
BoardFactory boardFactory = new CustomBoardFactory(snapShot, 16);
Board board = boardFactory.generate();
```

### **Q. 생성자 오버로딩(혹은 기존 정적 팩토리 메서드)에 비해 어떤 장점을 갖나요?**

> Board를 어떻게 만드는지에 대한 제어를 하고 싶으시다면 제어 하고싶은 영역의 의존성을 역전시켜보면 좋을것 같습니다.

이런 피드백을 주신 이유가, 의존성을 역전시키면 Board의 코드 수정 없이 생성 로직 변경이 가능하다는 이유라고 생각했습니다.
그러나 Board에는 코드 수정이 없겠지만, 대신 외부에서는 코드 수정이 발생이 불가피합니다. 변경되는 코드도 큰 차이가 없는 것 같습니다.

Board 생성자 오버로딩(혹은 기존 정적 팩토리 메서드)하여 생성하는 것과
```java
Board initialBoard = new Board(); // 초기화
Board customBoard = new Board(snapShot, turnCount); // 커스텀
```
의존성을 역전시켜 생성하는 것에 차이가 체감이 잘 안 됩니다.
```java
BoardFactory boardFactory = new InitialBoardFactory(); // 초기화
Board board = boardFactory.generate();
BoardFactory boardFactory = new CustomBoardFactory(snapShot, turnCount); // 커스텀
Board board = boardFactory.generate();
```
의존 역전 원칙(DIP)에 측면에서 의존성을 역전시키는 것이 좋다고 배웠지만, 해당 경우엔 어떤 장점이 있는지 잘 모르겠습니다.
제가 다르게 이해한 부분이 있다면 짚어주시면 감사드리겠습니다!!

pci2676 replied:
먼저 특정 방법이 옳다는것을 말씀 드리려는 것은 아닙니다.

위 상황에 국한하여 보았을때

특정 상황에서 커스텀한 정적 팩토리 메서드를 사용하시려는것으로 이해했습니다.

그리고 특정 상황은 테스트 상황으로 보였는데, 테스트 환경에서만 사용될 코드가 운영환경에 남아있는것은 좋지 않아 드린 리뷰였습니다.

테스트에서만 사용될 코드가 운영환경에 남아있지 않으면서 테스트하기 쉬운 환경을 조성할 수 있는 방법으로 의존성 역전을 말씀드린것이었습니다.

물론 상황이 변하여 운영환경에서도 정적팩토리 메서드 혹은 생성자의 오버로딩이 충분히 의미있게 쓰일 수 있게 된다면  
고안하신 방법이 더 좋아질 수 있을것 같습니다.

---
>대신 외부에서는 코드 수정이 발생이 불가피합니다.

위와 같이 말씀하신 부분은 제가 잘 이해하지 못했는데 더 자세하게 설명해주시면 좋을것 같습니다. 😄

eun-byeol replied:
`특정 상황에서 커스텀한 정적 팩토리 메서드를 사용하시려는것으로 이해했습니다.` 정확합니다😄
테스트에만 집중을 하다보니 놓친 포인트였던 것 같아요. 중요한 팁 감사합니다!👍

---

> 위와 같이 말씀하신 부분은 제가 잘 이해하지 못했는데 더 자세하게 설명해주시면 좋을것 같습니다.

Board를 호출하는 쪽에서 코드 수정이 발생한다는 의미로 말씀드렸습니다.
```java
BoardFactory boardFactory = new InitialBoardFactory(); // 초기화
Board board = boardFactory.generate();
BoardFactory boardFactory = new CustomBoardFactory(snapShot, turnCount); // 커스텀
Board board = boardFactory.generate();
```
위 코드에서 `new InitialBoardFactory()`, `new CustomBoardFactory(snapShot, turnCount)` 부분이 생성 전략에 따라 수정되는 부분이라고 생각했습니다.
그러나 이 부분은 의존성 역전과는 관련이 없는 질문이었던 것 같네요..하하🥲

---

> 테스트에서만 사용될 코드가 운영환경에 남아있지 않으면서 테스트하기 쉬운 환경을 조성할 수 있는 방법

수정된 코드 역시도 테스트 환경에 사용될 코드가 운영환경에 남아있는 문제가 있는 것 같습니다.
CustomBoardFactory를 test 폴더로 옮기고 사용한다면 해당 문제가 해소될 수 있을까요?
의도하신 `의존성 역전`을 제가 잘 구현한 것이 맞는지 궁금합니다.

pci2676 commented:
> ## 1. VO도 getter를 제한해야 하나요?
> 보드의 위치 좌표를 나타내는 Position VO 클래스가 있습니다. 위치 계산 로직을 위해 Position의 row와 column 값이 필요했습니다. VO를 위한 DTO 클래스를 만들어야 하는지, VO는 getter를 허용하는지 궁금합니다.

필요하면 노출해도 된다고 봅니다.

vo에게 메세지를 던져서 해결할수 있음에도 getter를 사용하여 값을 노출시키고 외부에서 연산을 수행하는 등의 행위를 해서 getter를 쓰지 말라고 많은 분들이 이야기 하는것으로 알고 있습니다.

pci2676 commented on src/main/java/chess/model/Position.java:
```diff
@@ -0,0 +1,63 @@
+
+    public static Position from(String coordinate) {
+        if (coordinate.length() != VALID_COORDINATE_LENGTH) {
+            throw new IllegalArgumentException("유효하지 않은 좌표값입니다.");
```
메세지에 원인을 추가해주시면 디버그에 용이할것 같습니다.

다른 부분도 같이 검토 부탁드립니다.

eun-byeol replied:
반영했습니다.
또한, move 로직을 Piece로 위임하면서, 기물 별로 예외를 던지도록 세분화하였습니다.

pci2676 commented on src/main/java/chess/model/Board.java:
```diff
@@ -0,0 +1,147 @@
+        board.put(target, sourcePiece);
+        board.put(source, Piece.from(PieceType.NONE));
+        turnCount++;
+    }
```
move에 private 메서드로 너무 많은 내용이 구현되어있는것 같습니다.

협력에 참여한 도메인 객체들에게 조금 더 책임을 부여해주시면 좋을것 같습니다.

eun-byeol replied:
board에서 수했던 경로 검증 로직을 piece 객체가 수행하도록 책임을 부여했습니다.
수정된 책임은 다음과 같습니다.
- Board : `position 유효성 검사`, `turn 검사`
- Piece : `기물 별 이동 방향 검사`, `경로 내 장애물 확인 검사`

단, piece에게 move 책임을 부여할 때, `모든 기물의 위치와 정보를 다 넘겨 주는 것이 괜찮을까?` 하는 의문이 들었습니다.
그러나, 해당 기물에서 경로 검사까지 하려면 모든 기물들을 알아야 하기 때문에 불가피했습니다.
```java
void move(Position source, Position target, Map<Position, Piece> pieces);
```
비밥은 해당 로직에 대해 어떻게 생각하시나요?

pci2676 commented on src/main/java/chess/model/Board.java:
```diff
@@ -0,0 +1,147 @@
+            if (targetPiece.isExist()) {
+                throw new IllegalArgumentException("경로 상에 다른 기물이 존재합니다.");
+            }
+        }
```
depth가 2를 넘어간 것 같습니다.

eun-byeol replied:
수정했습니다!!😄

pci2676 commented on src/main/java/chess/model/Board.java:
```diff
@@ -0,0 +1,147 @@
+        if (targetPiece.isEnemy(turnCount) && sourcePiece.isPawn() && ((Pawn) sourcePiece).canAttack(source, target)) {
+            return;
+        }
+        if (!sourcePiece.canMove(source, target)) {
```
`!`를 사용하여 반대되는 표현을 하는것은 코드를 읽을때 인지 과부하를 일으키는 요소로 작용할 수 있습니다.

가능한 한번에 읽고 이해할 수 있도록 조금 더 서술적인 표현을 메세지로 전달하는것을 권장드립니다.

eun-byeol replied:
`서술적인 표현` 유념하겠습니다!👍

pci2676 commented on src/main/java/chess/model/piece/Piece.java:
```diff
@@ -0,0 +1,93 @@
+    @Override
+    public String toString() {
+        return type.getDisplayName();
+    }
```
toString을 view를 위한 메서드로 사용하지 않는것이 좋습니다.

eun-byeol replied:
toString 메서드를 삭제했습니다.
`PieceMapper` enum 클래스를 생성하여, piece를 매핑해주는 방식으로 수정했습니다.

```java
public enum PieceMapper {

    PAWN("p", "P"),
    ROOK("r", "R"),
    ...

    public static String serialize(Piece piece) {
        PieceMapper pieceType = findPieceType(piece);
        Color color = Color.findColor(piece);
        if (color.isWhite()) {
            return pieceType.whitePiece;
        }
        return pieceType.blackPiece;
    }
}
```

pci2676 commented:
> ## 2. 로직의 통일성 vs 효율성
> 기물들의 이동 방향 검증을 Direction enum 클래스를 활용해서 구현하고자 했었습니다. 그러나, 일부 기물(Pawn, King, Knight)들은 Direction enum을 활용하는 것보다, 내부에서 따로 계산 로직을 만드는 것이 더욱 간결했습니다. 저는 효율성을 택하여 구현하는데, 이런 상황이라면 비밥은 어떤 것을 선택하실 지 궁금합니다🤔

일부 기물만 Direction을 사용하는 것이라면 따로 계산하는것도 조금은 이상한 모양새로 보입니다.

효율성과 통일성의 문제로 바라보는게 아닌 설계의 문제로 바라보시고 고민해보시면 좋을것 같습니다.

pci2676 commented on src/main/java/chess/model/piece/PieceType.java:
```diff
@@ -0,0 +1,60 @@
+    BLACK_KNIGHT("N", 1),
+    BLACK_BISHOP("B", 1),
+    BLACK_QUEEN("Q", 1),
+    BLACK_KING("K", 1),
```
도메인 모델이 view에 보여지는 속성까지 가지고 있는것은 view의 변화에 대응하기 힘들어보입니다.

기물의 색이 추가된다면 타입도 추가로 늘어나게될것 같습니다.

색상이라는 속성을 각 객체에 나타내는것이 유연한 설계일지 한번 고민해보시면 좋을것 같습니다.

eun-byeol replied:
> 도메인 모델이 view에 보여지는 속성까지 가지고 있는것은 view의 변화에 대응하기 힘들어보입니다.

도메인이 view를 모르게 하기 위해, `PieceMapper`에서 모든 기물의 변환 과정을 담당하도록 했습니다.
- serialize : Piece -> String
- deserialize : String -> Piece

> 색상이라는 속성을 각 객체에 나타내는것이 유연한 설계일지

타입과 색을 조합하여 기물을 생성하도록 수정했습니다.

pci2676 commented on src/test/java/chess/model/piece/PawnTest.java:
```diff
@@ -0,0 +1,100 @@
+        assertThat(canMove).isEqualTo(true);
+    }
+
+    @DisplayName("White Pawn이 전방 대각선 1칸 공격이면 canAttack true를 반환하고, 아니면 false를 반환한다")
```
테스트의 이름에는 메서드 이름을 노출시키지 않는것을 권장드립니다.

메서드 시그니처의 변경에 대처할 수 없습니다.

eun-byeol replied:
테스트 코드에서도 확장성을 고려할 수 있는 팁이네요! 감사합니다😊

pci2676 commented on src/test/java/chess/model/piece/RookTest.java:
```diff
@@ -0,0 +1,30 @@
+
+class RookTest {
+
+    @DisplayName("Rook이 상하좌우 이동이면 canMove true를 반환하고, 아니면 false를 반환한다")
```
```suggestion
    @DisplayName("Rook이 상하좌우 이동이면 움직일 수 있다.")
```

pci2676 commented on src/main/java/chess/controller/ChessGame.java:
```diff
@@ -0,0 +1,71 @@
+    private static final int TARGET_INDEX = 2;
+
+    public void run() {
+        InputView.printGameIntro();
```
view의 구현이 변경되어도 유연하게 대처할 수 있는 설계를 고려해보시면 좋을것 같습니다.

eun-byeol replied:
결론 : View 의존성을 주입받아서 사용하도록 수정했습니다.

`구현이 변경되어도 유연하게 대처할 수 있는 설계` 피드백을 저는 이렇게 이해했습니다. 올바르게 의도를 파악한 것이 맞을까요?
- 문제
  - 만약 ChessGame의 입력 View가 InputView가 아닌 다른 클래스를 사용하게 될 때, 기존의 코드에서는 InputView가 사용된 모든 메서드의 코드를 수정해줘야 한다.
- 해결 
  - ChessGame은 View를 멤버로 갖고, 생성 시 View의 의존성을 주입받으면, 위 상황에서 코드 수정이 불필요하다.


### **Q. 유연한 설계의 범위를 어디까지 설정하면 좋을까요?**
유연한 설계를 하는 것이 이상적이지만, 어디까지 유연하게 만들어야 할까에 대한 모호함이 있습니다.
미션 요구사항에 기반했을 때, View 클래스는 변하지 않는다는 스스로의 판단에 의해 Util 성격의 클래스를 만들어 사용했었습니다. 이런 관점에서는 정적 클래스가 합리적이라고 생각했었는데,
그러나 말씀해주신대로, View도 변할 수 있다고 본다면 의존성을 부여하는 것이 좋은 것 같습니다.

변할 수 있는지 혹은 변하지 않는 지, 어떻게 판단 기준을 세울 수 있을까요? 아직 설계에 대한 경험이 많이 부족하여, 이 부분에 스스로 답을 얻지 못했습니다. 이에 대해 조언을 구하고 싶습니다.🤔

pci2676 replied:
저의 경험상.. 경험의 영역에 가까운것 같습니다... 😅 

지금은 막연하실 수 있으나 시간이 지나면 많은 부분이 보이지 않을까 싶습니다.

pci2676 reviewed (changes requested):
반갑습니다. 🤠 

리뷰 확인부탁드리겠습니다.

pci2676 commented:
> 비밥의 시간규칙(?)을 알려주시면 좋을 것 같아요!(ex. 몇 시 이후에는 dm 자제 등)

DM은 언제든지 보내셔도 상관은 없는데 제가 즉답은 못해드릴 확률이 높습니다. 😅 

리뷰를 하는 시간은 딱히 정해져있지 않습니다만 보통 이른 아침이나 늦은 밤에 합니다. 😄

eun-byeol committed:
refactor: View static 키워드 삭제 및 컨트롤러에서 View 의존성 주입

eun-byeol committed:
test: DisplayName에 메서드명이 노출되지 않도록 수정

eun-byeol committed:
refactor: Factory 패턴 적용하여 체스판 생성 로직 수정

eun-byeol committed:
refactor: Board 멤버 변수명 수정

eun-byeol committed:
test: 커스텀 체스판 생성 테스트 추가

eun-byeol committed:
refactor: Position 예외 메시지 구체화

eun-byeol committed:
refactor: Piece 인스턴스 변수 수정, Mapper 생성

- Piece가 Type, Color 2개의 인스턴스 변수 갖도록 수정
- PieceMapper가 view와 model의 기물 데이터를 변환하는 역할 수행
- view를 위한 toString 메서드 삭제

eun-byeol committed:
style: 코드 들여쓰기 수정

eun-byeol committed:
refactor: Position 원시값 포장

eun-byeol committed:
refactor: move 로직 각 piece로 위임

- 경로 중 장애물 여부 검사 board에서 piece로 위임
- 기물 별 이동 조건 로직 수정 및 예외 메시지 추가
- Route 클래스 추가
- 테스트 Fixtures 추가

eun-byeol committed:
refactor: 패키지 변경, Direction 테스트 코드 추가

eun-byeol committed:
refactor: Route 인스턴스 변수 타입 변경 및 테스트 추가

eun-byeol committed:
test: Position 파라미터 테스트 케이스 추가

eun-byeol committed:
refactor: 게임 상태 관리 로직 객체화

- 컨트롤러에서 게임 종료 시 null 검증 로직 삭제
- GameStatus 클래스 생성하여 검증 위임

eun-byeol committed:
refactor: controller null 로직 삭제

- 게임 중 start는 1번만 수행하도록 강제

eun-byeol commented:
비밥, 안녕하세요!
깔끔하고 확장성이 좋은 설계를 하는 것이 정말 쉽지 않은 것 같습니다.🥲
앙파상, 캐슬링, 프로모션과 같은 특수 룰까지 적용하여 리뷰 요청을 드리고 싶었으나,
현재까지 구현된 코드를 많이 수정하게 되어 추가 기능 구현 없이 요청드리게 되었습니다.

부족함이 많은 코드이겠지만, 리뷰 주신 내용을 충족할 수 있는 방향으로 최대한 리팩터링 진행해 보았습니다.
이것 저것 시도해보고 적용해보느라, 커밋 단위가 깔끔하지 못한 점 양해 부탁드립니다..!

## 🤔 리팩토링 시 신경 쓴 점
### 1. Move 경로 검증 로직 Piece로 위임
> move에 private 메서드로 너무 많은 내용이 구현되어있는것 같습니다.
협력에 참여한 도메인 객체들에게 조금 더 책임을 부여해주시면 좋을것 같습니다.

경로 검증 로직을 Piece로 위임하여, Board가 Piece의 타입을 모르도록 수정했습니다.
수정 전 코드에서는 Board가 piece가 pawn인지, knight인지 확인하고, 분기하여 검증 로직을 수행하는 코드가 있었습니다.

수정 후 변경된 역할:
- Board : `position 유효성 검사`, `turn 검사`
- Piece : `기물 별 이동 방향 검사`, `경로 내 장애물 확인 검사`


### 2. 도메인 모델과 view를 분리하기 위해, Mapping 클래스 추가
> 도메인 모델이 view에 보여지는 속성까지 가지고 있는것은 view의 변화에 대응하기 힘들어보입니다.

도메인이 view를 모르게 하기 위해, `PieceMapper`에서 모든 기물의 변환 과정을 담당하도록 했습니다.
- serialize : Piece -> String
- deserialize : String -> Piece

또한, 기존에 모든 색깔 별 기물을 정의한 PieceType enum을 삭제하고, `Type`와 `Color` enum을 각각 조합하여 사용하도록 했습니다.

### 3. 기물 별 이동 Direction 검사 로직 통일
> 일부 기물만 Direction을 사용하는 것이라면 따로 계산하는것도 조금은 이상한 모양새로 보입니다.
효율성과 통일성의 문제로 바라보는게 아닌 설계의 문제로 바라보시고 고민해보시면 좋을것 같습니다.

pawn과 knight, king의 이동 로직도 Direction을 사용하도록 수정했습니다.
`source에서 target 위치가 갖는 direction을 구하고, 기물의 유효한 direction인지 확인`하도록 구현했습니다.

### 4. GameStatus 클래스에서 게임 상태 관리
> null을 다루는 코드보다 메세지를 활용하는 방식으로 개선해보시면 좋을것 같습니다.

- 문제 : controller에서 `board == null`이 될 때 game을 종료하는 로직
  - 해결 : GameStatus에서 게임 상태를 관리하도록 수정 (`READY` -> `START` -> `MOVE` -> `END`)
- 문제 : `board = null 로 초기화` 하는 로직
  - 해결 : 게임 시작 전 미리 체스판을 초기화하여 준비하고, start 명령어가 입력됐을 때 체스판 출력만 수행(start 명령어는 게임 중 1번만 수행되도록 강제)

최대한 메시지를 사용하도록 시도해보았는데,
`메시지를 활용하는 방식`에 대해 제가 올바르게 이해한 것이 맞을까요?

eun-byeol committed:
refactor: executeStart, executeMove 반환값 수정

pci2676 commented on src/main/java/chess/model/board/BoardFactory.java:
```diff
@@ -0,0 +1,35 @@
+            Piece piece = PieceMapper.deserialize(pieceName);
+            pieces.put(position, piece);
+        }
+    }
```
>수정된 코드 역시도 테스트 환경에 사용될 코드가 운영환경에 남아있는 문제가 있는 것 같습니다.
CustomBoardFactory를 test 폴더로 옮기고 사용한다면 해당 문제가 해소될 수 있을까요?
의도하신 의존성 역전을 제가 잘 구현한 것이 맞는지 궁금합니다.


의존성 역전은 잘 구현해주셨습니다.

default 메서드를 본래의 용도에 맞게 사용하고 있는지 검토해보시면 좋을것 같습니다.

pci2676 commented on src/main/java/chess/model/piece/MoveStrategy.java:
```diff
@@ -0,0 +1,9 @@
+public interface MoveStrategy {
+
+    void move(Position source, Position target, Map<Position, Piece> pieces);
+}
```
>단, piece에게 move 책임을 부여할 때, 모든 기물의 위치와 정보를 다 넘겨 주는 것이 괜찮을까? 하는 의문이 들었습니다.
그러나, 해당 기물에서 경로 검사까지 하려면 모든 기물들을 알아야 하기 때문에 불가피했습니다.

1.인터페이스화 하신 특별한 이유가 있을까요?

2.저는 3개의 매개변수를 사용하는 상황은 가능한 피하려고 하지만 불가피한 경우에는 어쩔 수 없을 것 같습니다.

move에 할당된 책임을 바깥으로 분리해보는것도 하나의 방법이 될 수 있을것 같습니다.

move에 부여하신 책임을 수행하기 위해 pieces를 억지로 넣은 모양새가 되는것 같다면 책임을 분리해보라는 신호로 한번 생각해볼 수 있을것 같습니다.

코드를 살펴보면 move에서 하는 일은 올바른 경로를 이동하는지를 검사하지 실제 이동을 하지 않는것으로 보였습니다.

move를 통해 이동경로를 반환하고 해당 이동경로에 대한 검사를 바깥(chessBoard)에서 객체간 메세지를 주고받으며 검사하는 방식으로도 해결해볼 수 있을것 같습니다.

반영을 하실필요는 없고 한번 고민해보시면 좋을것 같습니다.

eun-byeol replied:
1. 사실 인터페이스화 한 이유는 없었습니다. 
행동 규약을 인터페이스로 정의하고, 사용하는 객체에게 재정의를 강제하는 것이 가독성이 좋을 것이라고 생각했었습니다.
지금 상황에서는 인터페이스로 정의하고, Piece도 추상 클래스로 사용하고 있으니 불필요한 한 것 같습니다.
해당 인터페이스는 삭제했습니다!

2. 매개변수를 2개로 줄이는 방향으로 리팩터링 진행했습니다!
```java
public Route findRoute(Position source, Position target)
```

pci2676 commented on src/main/java/chess/model/piece/Pawn.java:
```diff
@@ -0,0 +1,79 @@
+        validateRoute(source, target, pieces);
+    }
+
+    public void validateAttack(Position source, Position target, Piece sourcePiece) {
```
올바르게 동작하지 않는것 같아보입니다.
```
start
move a2 a4
move b7 b5
move a4 b5
```
위 커맨드 수행시 `p`가 `P`를 공격하지 못하네요. 😢

eun-byeol replied:
테스트 코드를 더 꼼꼼하게 작성해야 하는 이유네요..! 감사합니다
수정 완료했고 테스트 케이스도 추가했습니다👍

pci2676 commented on src/main/java/chess/model/piece/Piece.java:
```diff
@@ -0,0 +1,81 @@
+            return new King(type, color);
+        }
+        return new None(type, color);
+    }
```
기물에 따라 if절이 나열되는 구현은 유지보수에 유리하지 않아보입니다.

가능한 단순화하게 구현할 수 있으면 좋을것 같습니다.

Piece의 구현체가 이미 Type을 나타내는것으로 보이는데 Type을 반드시 Piece가 알아야 하나요?

pci2676 commented on src/main/java/chess/controller/ChessGame.java:
```diff
@@ -0,0 +1,82 @@
+            return executeMove(commands, board, gameStatus);
+        }
+        return gameStatus;
+    }
```
>4. GameStatus 클래스에서 게임 상태 관리
null을 다루는 코드보다 메세지를 활용하는 방식으로 개선해보시면 좋을것 같습니다.
>
>문제 : controller에서 board == null이 될 때 game을 종료하는 로직
해결 : GameStatus에서 게임 상태를 관리하도록 수정 (READY -> START -> MOVE -> END)
문제 : board = null 로 초기화 하는 로직
해결 : 게임 시작 전 미리 체스판을 초기화하여 준비하고, start 명령어가 입력됐을 때 체스판 출력만 수행(start 명령어는 게임 중 1번만 수행되도록 강제)
최대한 메시지를 사용하도록 시도해보았는데,
메시지를 활용하는 방식에 대해 제가 올바르게 이해한 것이 맞을까요?

👍  null을 다루지 않도록 잘 해주신것 같습니다.

반드시 반영할 필요는 없지만 callback을 활용하시는것을 보아 아래 방법도 한번 고민해보시면 좋을것 같습니다.

지금 구현하신 내용에서 공통적인 부분만 추려보면 다음과 같이 보입니다.

- 확인하고 (isEnd, isStat, isMove, 3개 모두 아님)
- 행동하고 (changeEnd, executeStart, executeMove, gameStatus반환)

위 두가지 행위가 반복되는것을 GameStatus가 미리 알도록 하고 command만 넣었을때 지정된 행위를 하도록 해보는 것 입니다.

아래는 sudo 코드로 봐주세요

```java
var gameStatus = new GameStatus(
 onEnd = onEndCallback,
 onStart = onStartCallback,
 onMove = onMoveCallback
)

if(gameStatus.isRunning()){
  List<String> commands = inputView.askGameCommands();
  Command command = Command.findCommand(commands.get(COMMAND_INDEX));
  gameStatus = gameStatus.action(command)
}

```

eun-byeol replied:
함수형 인터페이스 적용하여 구현했습니다. 훨씬 깔끔한 것 같아요👍
```java
  GameStatus gameStatus = new GameStatus(this::executeStart, this::executeMove);

  List<String> commands = inputView.askGameCommands();
  gameStatus.action(commands, board);
```
콜백 함수를 객체화 하는 방식을 처음 본 것 같아요! 현업에서도 이런 패턴을 많이 사용하나요?

pci2676 replied:
사용해서 읽기 쉬워지는 코드가 작성된다면 사용하곤 합니다.

제가 있는 팀에서는 필요한 부분에 종종 사용하는 모습을 본 것 같네요.

pci2676 reviewed (changes requested):
질문 주신 부분에 대한 답변과 함께 리뷰를 남겨두었으니 확인해주세요.

eun-byeol committed:
fix: white pawn attack시 오른 대각선 이동 못 하는 오류 해결

eun-byeol committed:
refactor: Piece type 필드 삭제

eun-byeol committed:
refactor: Piece 생성 분기문 삭제

- PieceMapper에서 판단하도록 수정

eun-byeol committed:
refactor: 접근제어자 수정

eun-byeol committed:
refactor: GameStatus 로직 수정

eun-byeol committed:
refactor: move 로직 분리

- 기존 move 메서드 -> findRoute로 변경
- move 내에서 경로 검사하지 않고, 경로 반환 후 board에서 검증 수행
- pawn attack 검증 로직 메서드 분리
- Board move 시그니처 변경
- Board 인스턴스 변수 turnCount -> Color 타입으로 변경
- MoveStrategy 인터페이스 삭제

eun-byeol committed:
refactor: PiceMapper deserialize 로직 단순화

eun-byeol committed:
refactor: CustomBoardFactory test 디렉토리로 이동

eun-byeol committed:
refactor: 메서드 이름 변경

eun-byeol committed:
refactor: Movement로 클래스명 수정

- Direction -> Movement로 수정
- 메서드 이름 수정

eun-byeol committed:
refactor: 기물 생성 시 캐싱 적용

eun-byeol committed:
refactor: 접근제어자 private로 수정

eun-byeol committed:
style: 메서드 순서 변경

eun-byeol commented:
비밥, 안녕하세요!
리팩토링을 하면서 들었던 생각은.. 단순하게 할 수 있는데 왜 어렵게 생각했지? 이걸 왜 당연하게 생각했지? 하는 포인트가 많았던 것 같아요.
여러 질문을 던져주신 덕분에 많이 없앨 수 있었어요.
약간의 사담이었습니다..ㅎㅎ

반영하지 않아도 된다고 한 부분까지 전부 수정해 보았습니다.
사실 어떤 설계가 나을지 잘 모르겠어서 다 해보다보니, 반영하게 되었습니다.
리뷰 부탁드립니다🙂

## 🤔 리팩토링 시 신경 쓴 점
### 1. Piece type 필드 삭제
> Piece의 구현체가 이미 Type을 나타내는것으로 보이는데 Type을 반드시 Piece가 알아야 하나요?

Piece의 필드로 Color만 사용하도록 로직 수정했습니다.

단, `instanceOf`를 사용하지 않기 위해, Piece에 단순 반복 코드가 불필요하게 많이 생성됩니다.
```java
    public boolean isNone() {
        return false;
    }

    public boolean isPawn() {
        return false;
    }

    public boolean isRook() {
        return false;
    }

    public boolean isKnight() {
        return false;
    }

    public boolean isBishop() {
        return false;
    }

    public boolean isQueen() {
        return false;
    }

    public boolean isKing() {
        return false;
    }
```
가장 좋은 것은 기물의 타입 체크를 할 필요가 없는 설계를 하는 것인데, `PieceMapper`에서 Piece 객체 생성 시 사용이 불가피합니다.
```java
public enum PieceMapper {
    PAWN("p", "P", Pawn::of),
    ROOK("r", "R", Rook::of),
    KNIGHT("n", "N", Knight::of),
    BISHOP("b", "B", Bishop::of),
    QUEEN("q", "Q", Queen::of),
    KING("k", "K", King::of),
    NONE(".", ".", color -> None.of());
    ...
}
```
**Q. instanceOf를 사용하지 않고, 타입을 알 수 있는 방법이 있을까요?**

### 2. move 로직 분리
> move를 통해 이동경로를 반환하고 해당 이동경로에 대한 검사를 바깥(chessBoard)에서 객체간 메세지를 주고받으며 검사하는 방식으로도 해결해볼 수 있을것 같습니다.

기존 `move` 메서드를 `findRoute`로 변경하고, `이동 방향 검증` + `경로 반환` 까지만 수행하도록 역할을 축소했습니다.
장애물 유무 검증은 borad에서 수행하도록 수정했습니다.
```java
public Route findRoute(Position source, Position target)
```
또한, Pawn의 경우 `attack 검증 로직`(canAttack)을 move에서 별개의 메서드로 분리했습니다.
Board가 기물의 특성을 모르도록 구현하다보니
`Piece`에 Pawn만 사용하는 `canAttack`을 추가하게 되어, 다른 기물에 불필요한 메서드가 생겼습니다.

```java
public abstract class Piece {
    public boolean canAttack(Position source, Position target) {
        return false;
    }
    ...
}
```

**Q. Board에서 isPawn을 체크하지 않으면서, 다른 기물들에 불필요한 메서드가 생기지 않게 할 수 있는 다른 방안이 있을까요?**

### 3. 기물 생성 시 캐싱 적용
불필요한 객체 생성을 개선하기 위해, 캐싱을 적용했습니다.
```java
public class Pawn extends Piece {

    private static final Map<Color, Pawn> CACHE = Color.allColors()
        .stream()
        .collect(toMap(identity(), Pawn::new));

    public static Pawn of(Color color) {
        return CACHE.get(color);
    }
```

eun-byeol committed:
refactor: 인터페이스 추상 클래스로 변경

pci2676 commented:
>Q. instanceOf를 사용하지 않고, 타입을 알 수 있는 방법이 있을까요?

아래처럼 개선해볼 수 있을것 같습니다.

```java
    private static PieceMapper findPieceMapper(Piece piece) {
        Color color = Color.findColor(piece);
        return Arrays.stream(values())
                .filter(pieceMapper -> pieceMapper.generate.apply(color) == piece)
                .findFirst()
                .orElse(NONE);
    }
```

이렇게 되면 Type 자체가 필요없어질것 같아보이네요.

그러면 말씀하신 is~ 류의 메서드가 많이 제거될 것 같아보입니다.

pci2676 commented:
>Q. Board에서 isPawn을 체크하지 않으면서, 다른 기물들에 불필요한 메서드가 생기지 않게 할 수 있는 다른 방안이 있을까요?

공격와 움직임을 두가지로 분리해서 생각하고 계신것 같아보입니다.

공격도 하나의 움직임으로, 추상적으로 생각하고 구현해본다면 어떨까요?

pci2676 commented on src/main/java/chess/model/board/Board.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void move(Position source, Position target) {
+        validatePosition(source, target);
+        validateMovement(source, target);
+        applyMove(source, target);
```
어쩔수 없는 부분도 있기 마련이지만 가능한 행위에 대한 결과값을 반환하는 메서드를 작성해보시면 좋을것 같습니다.

테스트를 할 때 용이한 점도 있습니다. 상태에 대해 검사를 해야한다면 불필요한 getter가 필요해지는 경우도 발생할 수 있는데,

행위에 대한 검사는 결과 값 검사로 충분할 수 있기 때문입니다.

pci2676 commented on src/main/java/chess/model/board/Board.java:
```diff
@@ -0,0 +1,60 @@
+        turn = turn.rotate();
+    }
+
+    private void validatePosition(Position source, Position target) {
+        Piece sourcePiece = findPiece(source);
+        if (sourcePiece.isNone()) {
+            throw new IllegalArgumentException("source 위치에 기물이 존재하지 않습니다.");
+        }
+        if (sourcePiece.isDifferentColor(turn)) {
+            throw new IllegalArgumentException("지금은 " + turn.name() + " 차례입니다.");
+        }
+        if (findPiece(target).isAllyWith(sourcePiece)) {
+            throw new IllegalArgumentException("target 위치에 내 기물이 존재합니다.");
+        }
+    }
+
+    private void validateMovement(Position source, Position target) {
+        Piece sourcePiece = findPiece(source);
+        Piece targetPiece = findPiece(target);
+        if (sourcePiece.isEnemyWith(targetPiece) && sourcePiece.canAttack(source, target)) {
+            return;
+        }
+        Route route = sourcePiece.findRoute(source, target);
+        if (route.isBlocked(pieces)) {
+            throw new IllegalArgumentException("경로 상에 다른 기물이 존재합니다.");
+        }
+    }
```
이 부분도 조금 더 추상화할 수 있을지 고민해보시면 좋을것 같습니다.

구현된 코드의 추상적인 레벨이 조금 상이한것 같습니다.

리모콘을 만들어야 할 때 리모콘에
- 전원을 키고 끄는 기능과
- 채널 입력하는 기능을 

만들어달라고 했는데
만약 전원기능은 버튼 하나로 동작하는 반면 (버튼 하나로 추상화)
채널을 입력하는 기능이 회로에 전기신호를 직접 흘려야하는 구조가 노출되어있다고 생각해보시면 어떨까 싶습니다. (버튼으로 추상화되어있지 않고 드러나있음)

그래서 지금 구현한 코드에서 move를 하는데 만약 필요한 것이
움직일수 있는가에 대한 검사
기물을 교체하는것
두 가지라면 리모콘의 전원기능이 전원버튼 1개로 나타나듯 2개의 협력과정으로 나타나면 좋지 않을까요?

기존 협력객체를 이용해서 표현하기 힘들다면 새로운 협력객체가 등장해도 좋을것 같습니다.

pci2676 commented on src/main/java/chess/model/position/Movement.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.Arrays;
+
+public enum Movement {
+    UP(-1, 0),
+    DOWN(1, 0),
+    LEFT(0, -1),
+    RIGHT(0, 1),
+
+    UP_LEFT(-1, -1),
+    UP_RIGHT(-1, 1),
+    DOWN_LEFT(1, -1),
+    DOWN_RIGHT(1, 1),
+
+    UP_UP_LEFT(-2, -1),
+    UP_UP_RIGHT(-2, 1),
+    DOWN_DOWN_LEFT(2, -1),
+    DOWN_DOWN_RIGHT(2, 1),
+    LEFT_LEFT_UP(-1, -2),
+    LEFT_LEFT_DOWN(1, -2),
+    RIGHT_RIGHT_UP(-1, 2),
+    RIGHT_RIGHT_DOWN(1, 2),
+
+    UP_UP(-2, 0),
+    DOWN_DOWN(2, 0),
```
체스가 아닌 다른 게임을 구현한다고 가정해보았을때

그 게임에서는 기물이 움직일수 있는 방법이 무궁무진하다면 이렇게 enum으로 표현하기 힘들어질 수 있을것 같습니다.

그러한 부분을 해결할 수 있는 설계가 된다면 좋을것 같지만 구현 사항을 꽤 많이 고쳐야할 수 있기 때문에

반드시 반영하실 필요는 없고 한번 고민만 해보셔도 좋을것 같습니다.

pci2676 commented on src/test/java/chess/model/game/GameStatusTest.java:
```diff
@@ -0,0 +1,105 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class GameStatusTest {
```
상태를 유지하며 일련의 시나리오를 테스트하고 싶으시다면 

Junit의 DynamicTest를 활용해보시면 좋을것 같습니다.

지금처럼 상태를 공유하는 단위테스트는 지양하시는 것이 좋습니다.

아래 글의 2번항목을 읽어보시면 좋을것 같습니다.

https://blog.kingbbode.com/52

pci2676 reviewed (approved):
질문 남겨주신 부분을 비롯하여 조금 더 고민해보면 좋을 부분에 리뷰를 남겨두었습니다.

다음단계를 진행하며 같이 확인해주세요.

머지하겠습니다. 🚀
